File: src/main/java/org/vanautrui/languages/parsing/astnodes/nonterminal/statements/StatementNode.java
package org.vanautrui.languages.parsing.astnodes.nonterminal.statements;

import org.vanautrui.languages.lexing.collections.TokenList;
import org.vanautrui.languages.lexing.tokens.SymbolToken;
import org.vanautrui.languages.parsing.IASTNode;
import org.vanautrui.languages.parsing.astnodes.nonterminal.statements.controlflow.IfStatementNode;
import org.vanautrui.languages.parsing.astnodes.nonterminal.statements.controlflow.LoopStatementNode;
import org.vanautrui.languages.parsing.astnodes.nonterminal.statements.controlflow.ReturnStatementNode;
import org.vanautrui.languages.parsing.astnodes.nonterminal.statements.controlflow.WhileStatementNode;

public class StatementNode implements IASTNode {

    //TODO: add more statement types

    //can be method call ,loop statement, while statement, ...
    public IStatementNode statementNode;

    public StatementNode(TokenList tokens) throws Exception {
        //System.out.println("try parse DragonStatementNode");

        TokenList copy = tokens.copy();

        try {
            this.statementNode = new MethodCallNode(copy);
            copy.expectAndConsumeOtherWiseThrowException(new SymbolToken(";"));
        }catch (Exception e1){
            //maybe there is a loop statement
            try {
                this.statementNode = new LoopStatementNode(copy);
            }catch (Exception e2){
                try {
                    this.statementNode = new AssignmentStatementNode(copy);
                }catch (Exception e3){
                    try {
                        this.statementNode = new WhileStatementNode(copy);
                    }catch (Exception e4){
                        try {
                            this.statementNode = new IfStatementNode(copy);
                        }catch (Exception e5){
                            this.statementNode = new ReturnStatementNode(copy);
                        }
                    }
                }
            }
        }

        tokens.set(copy);
    }

    @Override
    public String toSourceCode() {
        return this.statementNode.toSourceCode();
    }
}

File: dragon-parser/src/main/compiler/parsing/astnodes/nonterminal/statements/StatementNode.cpp

//standard headers
#include <vector>

//project headers
#include "StatementNode.hpp"

public final class StatementNode implements IASTNode {

	//can be method call ,loop statement, while statement, ...
	public final IStatementNode statementNode;

	public StatementNode(final TokenList tokens) throws Exception {

		final TokenList copy = tokens.copy();

		if (copy.size() == 0) {
			throw new Exception("tried to parse a Statement, but there are no tokens left");
		}

		final IToken first = copy.get(0);

		if (first instanceof LoopToken) {
			this.statementNode = new LoopStatementNode(copy);
		} else if (first instanceof WhileToken) {
			this.statementNode = new WhileStatementNode(copy);
		} else if (first instanceof IfToken) {
			this.statementNode = new IfStatementNode(copy);
		} else if (first instanceof ReturnToken) {
			this.statementNode = new ReturnStatementNode(copy);
		} else {
			//TODO: we have to figure something out here.
			//i don't want 'let' statements
			//because it just messes up the code and is
			//just bloat to write.
			//but parsing should be straightforward. to give good error messages

			IStatementNode statementNode1; //temp variable so that this.statementNode can be final
			try {
				final TokenList copy2 = copy.copy();
				statementNode1 = new MethodCallNode(copy2);
				copy2.expectAndConsumeOtherWiseThrowException(new SemicolonToken());

				copy.set(copy2);
			} catch (Exception e1) {
				statementNode1 = new AssignmentStatementNode(copy);
			}
			this.statementNode = statementNode1;
		}

		tokens.set(copy);
	}

}
